xilinx/pack_dsp_xc7.cc

```diff
@@ -22,33 +23,136 @@
 
 NEXTPNR_NAMESPACE_BEGIN
 
+void XC7Packer::walk_dsp(CellInfo *current_cell, int constr_z)
+{
+    CellInfo *cascaded_cell = nullptr;
+
+    auto check_illegal_fanout = [&] (NetInfo *ni, std::string port) {
+        if (ni->users.size() > 1)
+            log_error("Port %s connected to net %s has more than one user", port.c_str(), ni->name.c_str(ctx));
+
+        PortRef& user = ni->users.back();
+        if (user.cell->type != ctx->id("DSP48E1_DSP48E1"))
+            log_error("User %s of net %s is not a DSP block, but %s",
+                user.cell->name.c_str(ctx), ni->name.c_str(ctx), user.cell->type.c_str(ctx));
+    };
+
+    // see if any cascade outputs are connected
+    for (auto port : current_cell->ports) {
+        if (!boost::contains(port.first.str(ctx), "COUT")) continue;
+        NetInfo *cout_net = port.second.net;
+
+        if (cout_net == nullptr) continue;
+
+        check_illegal_fanout(cout_net, port.first.c_str(ctx));
+        PortRef& user = cout_net->users.back();
+        CellInfo *cout_cell = user.cell;
+        NPNR_ASSERT(cout_cell != nullptr);
+
+        if (cascaded_cell != nullptr && cout_cell != cascaded_cell)
+            log_error("the cascading outputs of DSP block %s are connected to different cells",
+                current_cell->name.c_str(ctx));
+
+        cascaded_cell = cout_cell;
+    }
+
+    if (cascaded_cell != nullptr) {
+        cascaded_cell->constr_parent = current_cell;
+        current_cell->constr_children.push_back(cascaded_cell);
+        cascaded_cell->constr_x = 0;
+        // the connected cell has to be above the current cell,
+        // otherwise it cannot be routed, because the cascading ports
+        // are only connected to the DSP above
+        cascaded_cell->constr_y = constr_z == BEL_LOWER_DSP ? -5 : 0;
+        cascaded_cell->constr_z = constr_z;
+        cascaded_cell->constr_abs_z = true;
+
+        walk_dsp(cascaded_cell, constr_z == BEL_LOWER_DSP ? BEL_UPPER_DSP : BEL_LOWER_DSP);
```

Comment:
Can I ask you to set this up in a way so that all the DSPs in a cascade have the same `constr_parent` and increasing `constr_y` (i.e. there is only one `constr_parent` for a cascade instead of a linked-list type structure).

Otherwise this is going to break post rebase on master with the updated cluster API (sorry this got waylaid).

Reply:
Ok thanks for the feedback, will do that.

Reply:
OK, done that and tested working!